crates/sargon/src/system/sargon_os/transactions/sargon_os_transaction_analysis.rs

```diff
@@ -0,0 +1,705 @@
+    ) -> Result<TransactionPreviewResponse> {
+        let signer_public_keys =
+            self.extract_transaction_signers(manifest.summary()).await?;
+        let epoch = gateway_client.current_epoch().await?;
+        let header = TransactionHeader::new(
+            network_id,
+            epoch,
+            Epoch::from(epoch.0 + 10),
+            nonce,
+            notary_public_key,
+            signer_public_keys.is_empty(),
+            0,
+        );
+        let intent = TransactionIntent::new(header, manifest, message)?;
+        let request = TransactionPreviewRequest::new(
+            intent,
+            signer_public_keys,
+            TransactionPreviewRequestFlags::new(true, false, false),
+        );
+        let response = gateway_client.transaction_preview(request).await?;
+        if response.receipt.status != TransactionReceiptStatus::Succeeded {
+            return Err(Self::map_failed_transaction_preview(response));
+        };
+        Ok(response)
+    }
+
+    fn map_failed_transaction_preview(
+        response: TransactionPreviewResponse,
+    ) -> CommonError {
+        let message = response
+            .receipt
+            .error_message
+            .unwrap_or_else(|| "Unknown reason".to_string());
+
+        // Quite rudimentary, but it is not worth making something smarter,
+        // as the GW will provide in the future strongly typed errors
+        let is_failure_due_to_deposit_rules = message
+            .contains("AccountError(DepositIsDisallowed")
+            || message.contains("AccountError(NotAllBucketsCouldBeDeposited");
+
+        if is_failure_due_to_deposit_rules {
+            CommonError::OneOfReceivingAccountsDoesNotAllowDeposits
+        } else {
+            CommonError::FailedTransactionPreview {
+                error_message: message,
+            }
+        }
+    }
+
+    async fn extract_transaction_signers(
+        &self,
+        manifest_summary: ManifestSummary,
+    ) -> Result<IndexSet<PublicKey>> {
+        let signer_entities = self
+            .extract_transaction_signer_entities(manifest_summary.clone())
+            .await?;
+        Ok(self.extract_transaction_signers_public_keys(signer_entities))
+    }
+
+    async fn extract_transaction_signer_entities(
+        &self,
+        manifest_summary: ManifestSummary,
+    ) -> Result<IndexSet<AccountOrPersona>> {
+        let accounts =
+            self.profile_state_holder.accounts_on_current_network()?;
+        let personas =
+            self.profile_state_holder.personas_on_current_network()?;
+
+        let account_entities: IndexSet<_> = manifest_summary
+            .addresses_of_accounts_requiring_auth
+            .iter()
+            .filter_map(|address| {
+                accounts
+                    .iter()
+                    .find(|account| account.address == *address)
+                    .map(AccountOrPersona::AccountEntity)
+            })
+            .collect();
```

Comment:
put `IndexSet<_>` in the `collect` instead please. so `let variable_name = chain.chain.chain.collect::<TYPE>();`

Reply:
I think this **silences** non found entities,. right? we should use a `get_account(..) -> Result<Account>` and use:

```rust
addresses_of_accounts_requiring_auth
.into_iter()
.map(|a| self.profile_state_holder.get_account(a))
.collect::<Result<IndexSet<_>>>()?;
```

which will make this whole method return `Err` if any account was not found.

And analoguosly for Personas.